provider/core/src/any.rs

```diff
@@ -0,0 +1,386 @@
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use icu_provider::prelude::*;
+    /// use icu_provider::hello_world::*;
+    /// use std::borrow::Cow;
+    ///
+    /// const HELLO_DATA: HelloWorldV1<'static> = HelloWorldV1 {
+    ///     message: Cow::Borrowed("Custom Hello World")
+    /// };
+    ///
+    /// let any_payload = AnyPayload::from_static_ref(&HELLO_DATA);
+    ///
+    /// let payload: DataPayload<HelloWorldV1Marker> = any_payload.downcast()
+    ///     .expect("TypeId matches");
+    /// assert_eq!("Custom Hello World", payload.get().message);
+    /// ```
+    pub fn from_static_ref<Y>(static_ref: &'static Y) -> Self
+    where
+        Y: for<'a> Yokeable<'a>,
+    {
+        AnyPayload {
+            inner: AnyPayloadInner::StructRef(static_ref),
+            // Note: This records the Yokeable type rather than the DataMarker type,
+            // but that is okay since this is only for debugging
+            type_name: core::any::type_name::<Y>(),
+        }
+    }
+
+    /// Creates an `AnyPayload` from a [`DataPayload`] stored in an [`Rc`].
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use icu_provider::prelude::*;
+    /// use icu_provider::hello_world::*;
+    /// use std::borrow::Cow;
+    /// use std::rc::Rc;
+    ///
+    /// let payload: DataPayload<HelloWorldV1Marker> =
+    ///     DataPayload::from_owned(HelloWorldV1 {
+    ///         message: Cow::Borrowed("Custom Hello World")
+    ///     });
+    /// let rc_payload = Rc::from(payload);
+    ///
+    /// let any_payload = AnyPayload::from_rc_payload(rc_payload);
+    ///
+    /// let payload: DataPayload<HelloWorldV1Marker> = any_payload.downcast()
+    ///     .expect("TypeId matches");
+    /// assert_eq!("Custom Hello World", payload.get().message);
+    /// ```
+    pub fn from_rc_payload<M>(rc_payload: Rc<DataPayload<M>>) -> Self
+    where
+        M: DataMarker + 'static,
+    {
+        AnyPayload {
+            inner: AnyPayloadInner::PayloadRc(rc_payload),
+            type_name: core::any::type_name::<M>(),
+        }
+    }
+}
+
+impl<M> DataPayload<M>
+where
+    M: DataMarker + 'static,
+{
+    /// Wraps this DataPayload in an `Rc` and returns it as an `AnyPayload`.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use icu_provider::prelude::*;
+    /// use icu_provider::hello_world::*;
+    /// use std::borrow::Cow;
+    /// use std::rc::Rc;
+    ///
+    /// let payload: DataPayload<HelloWorldV1Marker> =
+    ///     DataPayload::from_owned(HelloWorldV1 {
+    ///         message: Cow::Borrowed("Custom Hello World")
+    ///     });
+    ///
+    /// let any_payload = payload.wrap_into_any_payload();
+    ///
+    /// let payload: DataPayload<HelloWorldV1Marker> = any_payload.downcast()
+    ///     .expect("TypeId matches");
+    /// assert_eq!("Custom Hello World", payload.get().message);
+    /// ```
+    pub fn wrap_into_any_payload(self) -> AnyPayload {
```

Comment:
thought: should we add a branch that avoids double wrapping? Or do you think that that's unnecessary or would cause bugs?

Reply:
A kind of bug I can see is when folks are writing generic code and expect double-wrapping

Reply:
The way I see it, every upcast (wrap) should be paired with a downcast, just as every malloc is paired with a free, and every open with a close.

I favor keeping the current behavior because:

1. I think it would be surprising if this function didn't guarantee the addition of a layer of wrapping, because then after you call it, you have an uncertain number of unwraps that you need to perform.
2. The purpose of this API is to help transform a typed DataProvider into an AnyProvider. Once you have an AnyProvider, there is no need to call this API. So I struggle to envision a scenario where a programmer would unwittingly perform a double wrapping.
3. The extra check and code path is going to be dead in most cases, but DCE probably won't be able to eliminate it most of the time.

Reply:
Seems reasonable!